File: src/main/java/meetme/conversations/ConversationsServiceImpl.java
package meetme.conversations;

import meetme.chats.ChatService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.simp.SimpMessageSendingOperations;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ConversationsServiceImpl implements ConversationsService {

    @Autowired
    ConversationsRepository conversationsRepository;

    @Autowired
    ChatService chatService;

    @Autowired
    private SimpMessageSendingOperations messagingTemplate;

    @Override
    public Conversations createChatMessage(Conversations conversations) {

//        conversations.setTimestamp(new Timestamp(new Date().getTime()));

        String threadId = conversations.getLoggedInUserId().compareTo(conversations.getOtherUserId()) > 0 ?
                conversations.getLoggedInUserId() + conversations.getOtherUserId() :
                conversations.getOtherUserId() + conversations.getLoggedInUserId();
        conversations.setThreadId(threadId);
        conversationsRepository.save(conversations);
        chatService.createChatMessage(conversations);

        //send WebSocket Message;
        this.messagingTemplate.convertAndSend("/conversations/" + conversations.getLoggedInUserId(), conversations);
        this.messagingTemplate.convertAndSend("/conversations/" + conversations.getOtherUserId(), conversations);
        this.messagingTemplate.convertAndSend("/chatThread/" + conversations.getThreadId(), conversations);

        return findConversationById(conversations.getLoggedInUserId()).get(0);

    }

    @Override
    public List<Conversations> findConversationById(String loggedInUserId) {
        return conversationsRepository.findByLoggedInUserId(loggedInUserId);
    }

    @Override
    public List<Conversations> findAllConversations() {
        return (List<Conversations>) conversationsRepository.findAll();
    }
}

File: src/main/java/meetme/conversations/ConversationsServiceImpl.kt
package meetme.conversations

import meetme.chats.ChatService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.messaging.simp.SimpMessageSendingOperations
import org.springframework.stereotype.Service

@Service
class ConversationsServiceImpl : ConversationsService {

    @Autowired
    lateinit var conversationsRepository: ConversationsRepository

    @Autowired
    lateinit var chatService: ChatService

    @Autowired
    lateinit var messagingTemplate: SimpMessageSendingOperations

    override fun createChatMessage(conversations: Conversations) {
        var conversation: Conversations = conversations
        conversation.threadId = getThreadId(conversations)
        conversation = saveConversation(conversations)
        val chat = chatService.createChatMessage(conversation)
        chatService.saveChatMessage(chat)

        this.messagingTemplate.convertAndSend("/conversations/" + conversations.firstUserId, conversation)
        this.messagingTemplate.convertAndSend("/conversations/" + conversations.secondUserId, conversation)
        this.messagingTemplate.convertAndSend("/chatThread/" + conversations.threadId, chat)

    }

    private fun saveConversation(conversation: Conversations): Conversations {
        val dbConversation: Conversations? = conversationsRepository.findById(conversation.threadId!!).orElse(null)
        conversation.unreadCounter = dbConversation?.unreadCounter?.inc() ?: 0
        return conversationsRepository.save(conversation)
    }

    override fun findConversationById(loggedInUserId: String): List<Conversations> {
        return conversationsRepository.findByFirstUserIdOrSecondUserId(loggedInUserId, loggedInUserId)
    }

    override fun findAllConversations(): List<Conversations> {
        return conversationsRepository.findAll() as List<Conversations>
    }

    fun getThreadId(conversations: Conversations) = if (conversations.firstUserId!!.compareTo(conversations.secondUserId!!) > 0)
        conversations.firstUserId + conversations.secondUserId else
        conversations.secondUserId + conversations.firstUserId
}
